### PR TITLE
Plotting - Automatic line coloring when plotting multiple expressions

### LINKED ISSUE #19384: Implement color cycler for plotting.
This is supported by matplotlib in their `plot` API, but sympy plotting bypasses that by using the the artist API, so there is a problem that multiple plots always having the same color by default.
If matplotlib provides some API for getting the color cycles, it can be mirrored here, however we can develop our own color cycler if that is not possible.

And the same issue should be investigated for different plotting APIs like contour or 3d plots, whether it cycles the colors or not.

```python3
import matplotlib.pyplot as plt
import numpy as np

x = np.linspace(-10, 10, 100)
y = np.cos(x)
y2 = np.sin(x)
b = plt.plot(x, y, x, y2, x, y)
```
image

```python3
from sympy import *

x = Symbol('x')
plot(cos(x), sin(x), (x, -10, 10))
```
image

### PR DESCRIPTION
**Pull request by Davide-sd**
#### Brief description of what is fixed or changed

1. Modified the Line series classes: now they return data which can
be used directly in matplotlib's plt.plot() method, which support
automatic line coloring. Previously, the format of the data was specific to
Matplotlib's `LineCollection`, which unfortunately doesn't support automatic
line coloring. The new format of the data makes it possible to easily create new
backends.
Back-compatibility has been maintained: it is still possible to apply a function to
the `line_color` attribute in order to create a color gradient.

2. Modified the `plot3d_parametric_line`, `plot3d`, `plot3d_parametric_surface`
methods in order to correctly set the axis labels in 3D plots.

3. Added the `zlabel` attribute to the `Plot` class, which will be rendered
by the backend. Similarly to a 2D plot where, for example, the x-label is set to
`"x"` and the y-label is set by default to `"f(x)"`, this edit will create a default
z-label `'"f(x, y)"` in 3D plots.

4. Modified the MatplotlibBackend class to:
	a. allow for automatic line coloring, while keeping back-compatibility
	with the `line_color` attribute.
	b. set the correct axis labels for 3D plots.

5. Updated the doc-strings associated to the `line_color` attribute.

#### References to other Issues or PRs

Fixes #19384

#### Release Notes

* plotting
  * Automatic line coloring

### CODE REVIEW

File: sympy/plotting/plot.py [hunk 1]
```diff
@@ -2289,6 +2358,9 @@ def plot3d_parametric_surface(*args, show=True, **kwargs):
     series = []
     plot_expr = check_arguments(args, 3, 2)
     series = [ParametricSurfaceSeries(*arg, **kwargs) for arg in plot_expr]
+    xlabel = kwargs.setdefault("xlabel", "x")
+    ylabel = kwargs.setdefault("ylabel", "y")
+    kwargs.setdefault("zlabel", "f(%s, %s)" % (xlabel, ylabel))
```
Comment (sylee957) [inline sympy/plotting/plot.py:2363]: For parametric line and surface, I think that it should better stay as `z`

Comment (Davide-sd) [reply sympy/plotting/plot.py]: I agree. Done it.

**Comment by sylee957** [comment]
I'm not sure why `legacy` keyword should be needed. Is it difficult to convert between list of segments vs list of points  by other means?

**Comment by Davide-sd** [comment]
Returning a list of segments is only required to use Matplotlib's `LineCollection`, which allows to use gradient coloring on the lines. So, the current implementation of `LineOver1DRangeSeries` and `Parametric2DLineSeries` is very Matplotlib specific. Note that by using `LineCollection` we get problems when we would like to show the legend, but that's another issue...

As far as I know, most of the other plotting libraries require list of points. Keep in mind that the user can implement their backends, PR #20463 ) (As a matter of fact, I usually use 3 different backeds depending on the problem at hand). Therefore, by using the `legacy` option, we make  `LineOver1DRangeSeries` and `Parametric2DLineSeries` almost backend-independent, hence reducing the code required to implement other backends (which is a very good thing, given Matplotlib limitations).

In my opinion, the series classes should only return list of points, and then `MatplotlibBackend` should implement its conversion function. Also, if it were for me, the `get_color_array` function of the `Line2DBaseSeries` class should also be moved into `MatplotlibBackend`. Unfortunately, that would break back-compatibility, that's why I created the `legacy` option.

**Comment by oscarbenjamin** [comment]
It would be better if sympy provided a function for getting the list of points in the first place and then the user could send them to whatever plotting library they like. Then sympy could provide high-level convenience routines for making some plots but there wouldn't be any need for a backend architecture because anyone who wanted something more complicated could easily do it themselves.

**Comment by Davide-sd** [comment]
Nope. The beauty of the current plotting module is the capability to create simple plots "independently" of the backend with just a single plot command. 

Sympy definitely needs new functions to easily convert a symbolic expression into numerical data that can be used by any plotting library, and part of that is already implemented in the Series classes. If the user find that any of the simple plot commands do not suit its needs, then it can use these functions and insert the numerical data into whatever library he wants.

**Comment by oscarbenjamin** [comment]
> with just a single plot command.

To be clear I don't object to having a single command like `plot(expr, use='plotly')`. What I object to is having that function return a wrapper class that tries to wrap the other libraries internally while hiding their native API.

**Comment by sylee957** [comment]
I think that it's better to update `get_points`  instead of modifying get_segments if you want to have a function to get points. 
`legacy` seems like an awkward keyword that indicates some deprecation need to be done, but doesn't follow https://github.com/sympy/sympy/wiki/Deprecating-policy

**Comment by Davide-sd** [comment]
> I think that it's better to update get_points instead of modifying get_segments if you want to have a function to get points.

That's easier said than done, because the actual `get_points` is used when `adaptive=False`. To do what you want it to do, it would require a substantial change in the current implementation of the Line Series classes, which would definitely break backcompatibility.

I think we can try the following:
* Move the adaptive algorithm implemented in `LineOver1DRangeSeries.get_segments` and `Parametric2DLineSeries.get_segments` into a new function `get_data`, which returns list of points.
* Remove the `legacy` option. Then, `get_segments` would first call `get_data` and convert the list of points to a list of segments. In doing so, backcompatibility should be maintained. Since this is a Matplotlib-specific feature, I would also raise a deprecation warning inside this function, and create an equivalent function in `MatplotlibBackend`.

In a later PR, the `get_data` function could also be implemented in the remaining Series classes. In doing so we can easily extract numerical data. For example, it should be possible to do:

```
# get numerical data to create custom plots starting from symbolic expressions
def plot_data(*args, **kwargs):
    args = list(map(sympify, args))
    # automatically create the series by looking at the form of the arguments and the number of free symbols
    # (this is possible, I've already tested it)
    s = _build_series(args, **kwargs)
    # simply call the common interface
    return s.get_data()
```

Without `get_data`, we would have to do something like:
```
# get numerical data to create custom plots starting from symbolic expressions
def plot_data(*args, **kwargs):
    args = list(map(sympify, args))
    # automatically create the series by looking at the form of the arguments and the number of free symbols
    # (this is possible, I've already tested it)
    s = _build_series(args, **kwargs)
    # simply call the common interface
    if s.is_implicit:
	return s.get_raster()
    elif s.is_3Dsurface:
	return s.get_meshes()
    ...
```

File: sympy/plotting/plot.py [hunk 2]
```diff
@@ -947,7 +1005,7 @@ def get_color_array(self):
             else:
                 return f(*variables)
         else:
-            return c*np.ones(self.nb_of_points)
+            return c*np.ones(min(self.nb_of_points_x, self.nb_of_points_y))
```
Comment (sylee957) [inline sympy/plotting/plot.py:1008]: Does this work with ParametricSurfaceSeries? I see it uses `nb_of_points_u, nb_of_points_v` instead of `x, y`

Comment (Davide-sd) [reply sympy/plotting/plot.py:1008]: Nope, need to fix it!

Comment (sylee957) [reply sympy/plotting/plot.py:1008]: I think that you should add some dry-run tests (without checking image output) in test_plot to see if these don’t raise some exceptions at least

Comment (Davide-sd) [reply sympy/plotting/plot.py]: Done. Thank you for pointing that out! :)

**Pull request merged**